app/jobs/kps_identity_update_job.rb

```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class KpsIdentityUpdateJob < ApplicationJob
+  queue_as :low
+
+  # slow operation
+  def perform(user, student_id = nil)
+    @user = user
+    @student_id = student_id
+    @response = Services::Kps::Omu::Kimlik.new.sorgula(user.id_number.to_i)
+  end
+
+  # callbacks
+  after_perform do |_job|
+    @user.identities.formal.update(@response.merge(student_id: @student_id))
+  end
+end
```

Comment:
`KpsIdentityCreateJob` sınıfıyla neredeyse aynı içeriğe sahip. Codebeat'den DRY uyarısı almışsın.

Reply:
Create ve Update için şimdilik yapılan işlemlerin hepsi aynı görünüyor. O zaman her iki işlemi yapan tek bir job yazabiliriz veya concern yapısını kullanmamız gerekiyor.

Tek bir job için örnek;

```ruby
# frozen_string_literal: true

class KpsIdentityCreateOrUpdateJob < ApplicationJob
  queue_as :high

  # slow operation
  def perform(user, student_id = nil)
    @user = user
    @student_id = student_id
    @response = Services::Kps::Omu::Kimlik.new.sorgula(user.id_number.to_i)
  end

  # callbacks
  after_perform do |_job|
    data = @response.merge(student_id: @student_id)
    formal = @user.identities.formal
    formal.present? ? formal.update(data) : formal.create(data)
  end
end
```

Reply:
Bazi seyleri netlestirmekte yarar var.

1) Kod tekrari her zaman kotu bir sey degil. Kod tekrarini onlemek adina yapacaginiz hareketler kodun okunurlugunu ciddi sekilde dusurecekse tekrar dusunmekte yarar var.

2) Codacy, codeclimate gibi servislerin uyarilari her zaman dogru olmayabilir. Bu gibi servisler yalnizca 'olasi' iyilestirmeleri oneriyor, bu oneriler sizi her zaman en iyi koda ulastirmayi falan garanti etmiyor. Zaten bu yuzden uyarilarda 'ignore', 'mute' gibi secenekler var. Kisacasi gelen her uyariyi sorgusuz sualsiz dogru kabul etmeyin.

Kod tekrarini azaltmak icin concern'e gelmeden once yapilabilecek baska hareketler de var. Ilki ve en basiti inheritance. Ikincisi ise module. Module'de bir nevi inheritance gibi zaten, Ruby'de single inheritance yapmak zorunda oldugumuz icin sikca module kullaniyoruz. Multiple inheritance yapabilseydik zaten mumkun mertebe hiyerarsik bir sinif tasarimiyla isleri cozme yoluna giderdik. Eger parent class'tan miras alma ve metodlari module'e alma da cozum olmuyorsa ya OOP'de birseyleri yanlis yapiyor, ya da tamamen Rails'e has bir sey yapiyor olmalisiniz.

Rails'e has durumlara Ruby bakis acisiyla bakmak da sakat, sonunda zorlama ve convention'dan uzak kodlar ortaya cikiyor.

Uzun lafin kisasi cok Rails'e has birseyle karsilastiysaniz, inheritance ve module isleri toparlamadiysa son care concern'e bakin - bir kac satir tekrar etti diye kodun okunurlugunu feda ederek sizden sonra gelecek gelistiriciye sac bas yoldurtmayin.

Reply:
@msdundar :+1: (özellikle "Rails way" kısmına).  Bir ekleme de ben yapayım.  Kod tekrarını çözmek bazen kolay, bazen de (okunurluğu azaltmayacak şekilde) çok zor.  Bu ikinci sınıf "zor" kod tekrarı (tecrübeme göre) çoğunlukla tasarımdaki bir "flaw"a işaret eder.  Altın kurallar:

- DRY uyarılarını mutlaka dikkate alın ve **çözmeye çalışın**.  Bu bir süreç, sürecin sonunda mutlaka çözmüş olmanız gerekmiyor, süreci yaşamanız gerekiyor.

- Sadece DRY yapmak adına koda **erken müdahalelerde bulunmayın**. Özellikle erken müdahale erken soyutlamalara (ve okunması zor kodlara) yol açacaksa.  Vaktiniz varsa tasarımı yeni baştan yapın, vaktiniz yoksa kod tekrarını göz ardı edip başka (ve geniş) bir zamana bırakın bu işi.

@isubas Önerdiğin düzeltme bence gayet yerinde.  (Bana göre) tek sıkıntı "CreateOrUpdate" gibi kimliksiz ve "Do one thing and do it well" ilkesini ihlal ettiği izlenimini veren bir isim kullanılması. Bu aslında sıklıkla karşılaşılan bir isimlendirme sorunu.  Önerilere baktım: [1](https://english.stackexchange.com/questions/227915/word-meaning-both-create-and-update), [2](https://stackoverflow.com/questions/33418898/what-is-the-correct-http-method-for-a-create-or-update-action-in-a-rails-api).  "Save" uygun bir karşılık gibi duruyor (Rest bağlamında "Put" daha uygun olurdu).  Böyle uzun yazıyorum ki bir konvansiyon oluşsun: `KpsIdentityCreateOrUpdateJob` yerine `KpsIdentitySaveJob` dediğimizde buradaki `Save`'in "create or update" semantiğinde kullanıldığını herkes anlasın ve bundan sonra da benzer durumlarda kullanılsın diye.  Save dışında bir seçenek de olabilir, yeter ki üzerinde anlaşmaya bağlı bir konvansiyon oluşsun.